pkg/apis/networking/v1alpha1/clusteringress_validation.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+
+	"github.com/knative/pkg/apis"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func (ci *ClusterIngress) Validate() *apis.FieldError {
+	return ci.Spec.Validate().ViaField("spec")
+}
+
+func (spec *IngressSpec) Validate() *apis.FieldError {
+	fmt.Printf("%#v\n", spec)
+	// Spec must not be empty.
+	if equality.Semantic.DeepEqual(spec, &IngressSpec{}) {
+		return apis.ErrMissingField(apis.CurrentField)
+	}
+	// Spec must have at least one rule.
+	if len(spec.Rules) == 0 {
+		return apis.ErrMissingField("rules")
+	}
+	// Validate each rule.
+	for idx, rule := range spec.Rules {
+		if err := rule.Validate(); err != nil {
+			return err.ViaField(fmt.Sprintf("rules[%d]", idx))
```

Comment:
Consider collecting all the errors like so:

```golang
err := (*apis.FieldError)nil
for (....) {
  err = err.Also(err.ViaField(fmt.Sprintf("rules[%d]", idx)))
}
for (...) {
  err = err.Also(err.ViaField(fmt.Sprintf("tls[%d]", idx)))
}
return err
```

`Also` will return a `nil` error if the bound object and the arguments are all `nil`.